pkg/querier/query_frontend_queryable.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/thanos-io/thanos/pkg/promclient"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/querier/series"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+const orgIDHeader = "X-Scope-OrgId"
+
+type queryFrontendQueryable struct {
+	step       time.Duration
+	promClient *promclient.Client
+	endpoint   *url.URL
+}
+
+type queryFrontendQuerier struct {
+	mint, maxt int64
+	step       int64
+
+	promClient *promclient.Client
+	endpoint   *url.URL
+}
+
+func NewQFQueryable(cfg Config, frontendAddress string, reg prometheus.Registerer, logger log.Logger) (storage.SampleAndChunkQueryable, promql.QueryEngine) {
+	qfUrl, _ := url.Parse(frontendAddress)
+	queryable := NewQueryFrontendQueryable(qfUrl, logger)
+
+	var queryEngine promql.QueryEngine
+	opts := promql.EngineOpts{
+		Logger:               logger,
+		Reg:                  reg,
+		ActiveQueryTracker:   createActiveQueryTracker(cfg, logger),
+		MaxSamples:           cfg.MaxSamples,
+		Timeout:              cfg.Timeout,
+		LookbackDelta:        cfg.LookbackDelta,
+		EnablePerStepStats:   cfg.EnablePerStepStats,
+		EnableAtModifier:     true,
+		EnableNegativeOffset: true,
+		NoStepSubqueryIntervalFn: func(int64) int64 {
+			return cfg.DefaultEvaluationInterval.Milliseconds()
+		},
+	}
+	if cfg.ThanosEngine {
+		queryEngine = engine.New(engine.Opts{
+			EngineOpts:        opts,
+			LogicalOptimizers: logicalplan.AllOptimizers,
+		})
+	} else {
+		queryEngine = promql.NewEngine(opts)
+	}
+
+	return NewSampleAndChunkQueryable(queryable), queryEngine
+}
+
+func NewQueryFrontendQueryable(url *url.URL, logger log.Logger) storage.Queryable {
+	httpClient := &http.Client{
+		Timeout: time.Second * 2,
+	}
+
+	return &queryFrontendQueryable{
+		step:       time.Second * 60,
+		promClient: promclient.NewClient(httpClient, logger, "cortex-ruler"),
+		endpoint:   url,
+	}
+}
+
+func (q *queryFrontendQueryable) Querier(mint, maxt int64) (storage.Querier, error) {
+	return &queryFrontendQuerier{
+		mint:       mint,
+		maxt:       maxt,
+		step:       int64(q.step / time.Second),
+		promClient: q.promClient,
+		endpoint:   q.endpoint,
+	}, nil
+}
+
+// Select implements storage.Querier interface.
+func (q *queryFrontendQuerier) Select(ctx context.Context, _ bool, _ *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+	log, ctx := spanlogger.New(ctx, "queryFrontendQuerier.Select")
+	defer log.Span.Finish()
+
+	query := storepb.PromMatchersToString(matchers...)
+
+	//lint:ignore faillint wrapper around upstream method
+	orgId, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		level.Error(log).Log("err", err, "orgId", orgId)
+		return storage.ErrSeriesSet(err)
+	}
+
+	m, warns, _, err := q.promClient.QueryRange(ctx, q.endpoint, query, q.mint, q.maxt, q.step, promclient.QueryOptions{
+		HTTPHeaders: http.Header{orgIDHeader: []string{orgId}},
+		Deduplicate: true,
+	})
```

Comment:
Can we avoid using Thanos PromClient?

Reply:
Thank you for the guidance. I going to fix it.